src/ucp/core/ucp_ep.c

```diff
@@ -983,16 +987,19 @@ ucs_status_ptr_t ucp_ep_close_nb(ucp_ep_h ep, unsigned mode)
 void ucp_ep_discard_lanes(ucp_ep_h ep, ucs_status_t status)
 {
     ucp_lane_index_t lane;
+    uct_ep_h uct_ep;
 
     for (lane = 0; lane < ucp_ep_num_lanes(ep); ++lane) {
-        if (ep->uct_eps[lane] == NULL) {
+        uct_ep = ep->uct_eps[lane];
+
+        ucp_ep_reset_lane(ep, lane);
+
+        if (uct_ep == NULL) {
             continue;
         }
 
-        ucs_debug("ep %p: discard uct_ep[%d]=%p", ep, lane,
-                  ep->uct_eps[lane]);
-        ucp_worker_discard_uct_ep(ep, ep->uct_eps[lane],
-                                  UCT_FLUSH_FLAG_CANCEL,
+        ucs_debug("ep %p: discard uct_ep[%d]=%p", ep, lane, uct_ep);
+        ucp_worker_discard_uct_ep(ep, uct_ep, UCT_FLUSH_FLAG_CANCEL,
```

Comment:
should we also set all lanes to failed before doing discard?
in case discard will call flush that will call ucp request callback that will access another lane?
if yes lets have common helper fuinction for that

Reply:
yes, better to do it

Reply:
done